src/fastpath/fastpath.c

```diff
@@ -6,15 +6,175 @@
+    fault_type = seL4_Fault_get_seL4_FaultType(NODE_STATE(ksCurThread)->tcbFault);
+
+    /* Check there's no saved fault */
+    if (unlikely(fault_type != seL4_Fault_NullFault)) {
+        slowpath(SysSend);
+    }
+
+    /* Lookup the cap */
+    cap_t cap = lookup_fp(TCB_PTR_CTE_PTR(NODE_STATE(ksCurThread), tcbCTable)->cap, cptr);
+
+    /* Check it's a notification */
+    if (unlikely(!cap_capType_equals(cap, cap_notification_cap))) {
+        slowpath(SysSend);
+    }
+
+    /* Check that we are allowed to send to this cap */
+    if (unlikely(!cap_notification_cap_get_capNtfnCanSend(cap))) {
+        slowpath(SysSend);
+    }
+
+    /* Get the notification address */
+    notification_t *ntfnPtr = NTFN_PTR(cap_notification_cap_get_capNtfnPtr(cap));
+
+    /* Get the notification state */
+    uint32_t ntfnState = notification_ptr_get_state(ntfnPtr);
+    tcb_t *dest;
+
+    if (ntfnState == NtfnState_Waiting) {
+        /* get the destination thread */
+        dest = TCB_PTR(notification_ptr_get_ntfnQueue_head(ntfnPtr));
+    } else {
+        /* get the bound tcb */
+        dest = (tcb_t *) notification_ptr_get_ntfnBoundTCB(ntfnPtr);
+    }
+
+    /* Get the notification badge */
+    word_t badge = cap_notification_cap_get_capNtfnBadge(cap);
+    switch (ntfnState) {
+    case NtfnState_Active: {
+#ifdef CONFIG_BENCHMARK_TRACK_KERNEL_ENTRIES
+        ksKernelEntry.is_fastpath = true;
+#endif
+        ntfn_set_active(ntfnPtr, badge | notification_ptr_get_ntfnMsgIdentifier(ntfnPtr));
+        restore_user_context();
+        UNREACHABLE();
+    }
+    case NtfnState_Idle: {
+        /* Check if we are bound and that thread is waiting for a message */
+        if (dest && thread_state_ptr_get_tsType(&dest->tcbState) == ThreadState_BlockedOnReceive) {
+            /* Signal to higher prio thread is NOT fastpathed*/
+            if (NODE_STATE(ksCurThread)->tcbPriority < dest->tcbPriority) {
+                slowpath(SysSend);
+            }
+
+            /* Basically equivalent to maybeDonateSchedContext. Check whether the thread already has
+             * a SC  or if one can be donated from the notification. If neither is true, go to
+             * slowpath */
+            if (!dest->tcbSchedContext) {
+                sc = SC_PTR(notification_ptr_get_ntfnSchedContext(ntfnPtr));
+                if (sc == NULL || sc->scTcb != NULL) {
+                    slowpath(SysSend);
+                    UNREACHABLE();
+                }
+            }
+
+            /*  Point of no return */
+
+            /* Equivalent to cancel_ipc */
+            endpoint_t *ep_ptr;
+            ep_ptr = EP_PTR(thread_state_get_blockingObject(dest->tcbState));
+            endpoint_ptr_set_epQueue_head_np(ep_ptr, TCB_REF(dest->tcbEPNext));
+            if (unlikely(dest->tcbEPNext)) {
+                dest->tcbEPNext->tcbEPPrev = NULL;
+            } else {
+                endpoint_ptr_mset_epQueue_tail_state(ep_ptr, 0, EPState_Idle);
+            }
```

Comment:
I don't think this code handles the case where the TCB isn't at the head of the endpoint queue. `cancel_ipc` calls `tcbEPDequeue()` which only updates the head of the queue of the TCB is at the front. This code unconditionally updates the head of the queue because it assumes that the TCB is always at the front. Similarly with the tail.